keps/sig-node/4328-node-label-manager/README.md

```diff
@@ -0,0 +1,442 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+This KEP aims to implement the `RequiredDuringSchedulingRequiredDuringExecution` type for NodeAffinity. This type implies that pods with this affinity rule must meet certain conditions during both the scheduling phase and the execution phase. 
+
+## Motivation
+- Enable strict node affinity for pod scheduling and execution.
+- Provide users with greater control over pod placement.
+- Ensure predictable and consistent pod behavior in various scenarios.
+- Address use cases with specific placement and execution requirements.
+  
+### Goals
+- Introduce the "RequiredDuringSchedulingRequiredDuringExecution" node affinity type.
+- Add a node affinity controller to the controllers to ensure pods being evicted if the selectors are no longer met at runtime.
+
+### Non-Goals
+- This KEP does not affect existing node affinity types.
+- It does not introduce breaking changes for workloads with different requirements.
+- It does not enforce stricter affinity rules for existing workloads.
+- It does not modify core Kubernetes components unrelated to node affinity.
+
+## Proposal
+- Implement the "RequiredDuringSchedulingRequiredDuringExecution" node affinity type.
+- Update the Kubernetes scheduler and relevant components.
+- Define the behavior, syntax, and use cases in documentation.
+- Facilitate testing, adoption, and maintenance of this feature.
+
+### User Stories (Optional)
+
+#### Optimizing Workloads Migrating
+As a Data Scientist, I want to ensure that my GPU-intensive ml workloads are always on nodes with correct type of GPUs during both scheduling and execution. So I labeled one node with "gpu=g1". I added some more powerful nodes, and now I want to move all pods requrieing gpu to this new node. With this feature, I can simply remove the "gpu=g1" from the first node and adds it to the new node, so I can move the target pods and keep the remaining pods intact.
+
+### Notes/Constraints/Caveats (Optional)
+
+N/A
+
+### Risks and Mitigations
+
+The major concern for this controller may be the overhead to the nodes with a large number of pods. 
+Every time a node's label is updated, the controller needs to iterate all pods assigned to the node. This may cause potential scaling issues.
+
+## Design Details
+
+Add a node label manager similar to the existing taint_eviction controller, but under the node lifecycle controller.
+
+The manager do the following things:
+
+- Listening to the changes of node labels
+- Iterating over all pods assigned to the nodes, check the NodeAffinity field, if `RequiredDuringSchedulingRequiredDuringExecution` exists, check if the selector still met to the new node label. If not, evict the pod.
+
+As for API definition, `RequiredDuringSchedulingRequiredDuringExecution` will have the same item as `requiredDuringSchedulingIgnoredDuringExecution`.
```

Comment:
I would maybe include the API addition here. What type/name are you going to add to the pod api?

Reply:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L3211 I guess is what you are enabling?

Reply:
Added. 
https://github.com/kubernetes/kubernetes/blob/8254569d50d6bbe8d60bcdfb16eac337f6e60514/staging/src/k8s.io/api/core/v1/types.go#L3354 is what I'm enabling.